### ISSUE TITLE
osproc: execProcess and startProcess hang forever (stuck on streams.readLine)

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#9953]
I'm on OSX

any `osproc` routine that captures stdout gets stuck forever.

nim c -r -d:case1 $timn_D/tests/nim/all/t0040.nim
```
before
^CTraceback (most recent call last)
/Users/timothee/git_clone/nim/timn/tests/nim/all/t0040.nim(29) t0040
/Users/timothee/git_clone/nim/timn/tests/nim/all/t0040.nim(23) main
/Users/timothee/git_clone/nim/Nim/lib/pure/osproc.nim(365) execProcess
/Users/timothee/git_clone/nim/Nim/lib/pure/streams.nim(280) readLine
/Users/timothee/git_clone/nim/Nim/lib/pure/streams.nim(162) readChar
/Users/timothee/git_clone/nim/Nim/lib/pure/streams.nim(87) readData
/Users/timothee/git_clone/nim/Nim/lib/pure/streams.nim(413) fsReadData
SIGINT: Interrupted by Ctrl-C.
```

nim c -r -d:case2 $timn_D/tests/nim/all/t0040.nim
```
before
^CSIGINT: Interrupted by Ctrl-C.
Traceback (most recent call last)
/Users/timothee/git_clone/nim/timn/tests/nim/all/t0040.nim(29) t0040
/Users/timothee/git_clone/nim/timn/tests/nim/all/t0040.nim(27) main
/Users/timothee/git_clone/nim/timn/tests/nim/all/t0040.nim(16) startProcessWrapper
/Users/timothee/git_clone/nim/Nim/lib/pure/streams.nim(306) readLine
/Users/timothee/git_clone/nim/Nim/lib/pure/streams.nim(75) atEnd
/Users/timothee/git_clone/nim/Nim/lib/pure/streams.nim(408) fsAtEnd
/Users/timothee/git_clone/nim/Nim/lib/system/sysio.nim(233) endOfFile
SIGINT: Interrupted by Ctrl-C.
```

```nim
import osproc
import streams

let cmd = "/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"

proc startProcessWrapper()=
  var p = startProcess(command = cmd, options={poStdErrToStdOut})
  defer: close(p)
  let sub = outputStream(p)
  while true:
    echo "before"
    let line = readLine(sub)
    echo line
    echo "after"

proc main()=
  when defined(case1):
    echo "before"
    let ret = execProcess(cmd, options={poStdErrToStdOut})
    echo "after"
    echo ret
  when defined(case2):
    startProcessWrapper()

main()
```

## other languages don't have this problem
in D:
```d
import std.process;
import std.stdio;

void main(){
  auto ret = execute("/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl");
  writeln("ret:", ret);
}
```

```
rdmd $timn_D/tests/nim/all/t0039.d
ret:Tuple!(int, "status", string, "output")(0, "")
```

**Comment by alaviss** [comment]
Do you have any easily reproducible sample that only requires POSIX tools? Not all of us have Sublime Text to test this with.

**Comment by Araq** [comment]
These APIs are heavily used. By testament, on OSX too. Nothing hangs. Maybe it's the space in your path?

**Comment by timotheecour** [comment]
i'm not making it up :)
not a space in path issue, same thing after `cp "/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl" /tmp/z01` and trying on `/tmp/z01`

* some progress though: 
the issue is inside `startProcess`, because if I call C code on 
`var p = startProcess(command = cmd, options={poStdErrToStdOut, poEvalCommand})` (with proper casting , I know what I'm doing here) I will get a hang depending on what command I give.

* furthermore, the issue is with buffering: 
indeed, everything works If my command is `unbuffer /tmp/z01` instead of `/tmp/z01`

**Comment by timotheecour** [comment]
@alaviss > Do you have any easily reproducible sample that only requires POSIX tools? Not all of us have Sublime Text to test this with.

not yet but pretty sure it's not specific to `sublime text` ; it's got to do with stdout buffering and whether application returns no output at all if I were to guess

**Comment by timotheecour** [comment]
* other finding:
`subl` returns on cmd line (after opening a sublime window)
but `echo yes | /tmp/z01` doesn't return on cmd line, it opens a sublime buffer called `subl stdin odEsPn.txt` (presumably `subl stdin <random file>.txt`), writes what you typed to stdin (yes) in that window, and waits for user to close the window, and finally returns on cmd line.

that could explain the bug ; it's still a bug IMO as both D and C++ work in that case.

**Comment by alaviss** [comment]
Here's my speculation: `startProcess` creates a pipe between the child and the parent, which causes the child's `stdin` to no longer be connected to the console, which causes sublime to enter the `subl stdin` buffer.

I think a partial version of `poParentStreams` that forwards only the `stdin` may work.

**Comment by alaviss** [comment]
Here's my PoC:

`child.nim`:
```nim
import terminal

if isatty stdin:
  echo "Hi!"
else:
  while true: discard
```

`parent.nim`:
```nim
import osproc

let ret = execProcess("./child", options={poStdErrToStdOut})
```

This will hang indefinitely unless `poParentStreams` is passed to `execProcess`

**Comment by timotheecour** [comment]
@alaviss thanks for the reply; but passing `poParentStreams` crashes:

/Users/timothee/git_clone/nim/Nim/lib/pure/osproc.nim(360) execProcess
/Users/timothee/git_clone/nim/Nim/lib/pure/osproc.nim(1251) outputStream
/Users/timothee/git_clone/nim/Nim/lib/system.nim(3930) failedAssertImpl
/Users/timothee/git_clone/nim/Nim/lib/system.nim(3923) raiseAssert
/Users/timothee/git_clone/nim/Nim/lib/system.nim(2969) sysFatal
Error: unhandled exception: /Users/timothee/git_clone/nim/Nim/lib/pure/osproc.nim(1251, 17) `poParentStreams notin p.options` API usage error: stream access not allowed when you use poParentStreams [AssertionError]
Error: execution of an external program failed: '/Users/timothee/git_clone/nim/timn/tests/nim/all/t0044 '

**Comment by alaviss** [comment]
Yes, that's why I propose a `poParentStdin` instead to allow forwarding only stdin. `poParentStreams` will forward both stdin and stdout, so you can't capture the output.

**Comment by timotheecour** [comment]
@alaviss thanks a bunch for your help!, that seems to fix this bug, see https://github.com/nim-lang/Nim/pull/9963 for WIP

**Comment by genotrance** [comment]
Not sure if https://github.com/nim-lang/Nim/issues/8648 is related.

**Comment by zah** [comment]
I agree the process API needs to be reworked. I've filed a similar issue here:
https://github.com/nim-lang/Nim/issues/7999

**Comment by trialism** [comment]
> These APIs are heavily used. By testament, on OSX too. Nothing hangs. Maybe it's the space in your path?

I have this problem too: [on nimforum](https://forum.nim-lang.org/t/4454).

**Comment by davidgarland** [comment]
Is there any workaround or alternative library for getting a setup like this to work currently? Or are we stuck until the API is reworked?

**Comment by timotheecour** [comment]
I've solved this issue and other problems related to osproc (eg https://github.com/nim-lang/Nim/issues/7999) by doing a full rewrite inspired from D's `std/process.d` approach. I don't have time/energy for a PR on this at the moment (notably, only works on linux/osx, didn't test on windows) but can share a bit more details if anyone is interested.
The API is flexible, allowing one to redirect stdin/stdout/stderr independently (as opposed to all or nothing `poParentStreams`) to a file, pipe or other descriptor, along with other improvements.

**Comment by alaviss** [comment]
So what is the API that you've arrived at? I'm interested :)

**Comment by zah** [comment]
@timotheecour, would you post your new code as a github repo? Perhaps the community can help to develop the Windows version and to integrate it with the async libraries. I'd like to have good process management support in our [Chronos](https://github.com/status-im/nim-chronos/) library for example.

**Comment by alehander42** [comment]
i want independent parent streams

**Comment by zevv** [comment]
Changing the API would be the best solution, but in the mean time this could be fixed by using temp files?

**Comment by zevv** [comment]
I recently verified this also affects the Nim compiler itself: https://github.com/nim-lang/Nim/issues/8648

**Comment by hoijui** [comment]
I wrote a tool that makes heavy use of `osproc`, and this makes it unusable. I am considering switching the whole software to rust because of it.
Please, fix this!

**Comment by hoijui** [comment]
I am writing a CLI application that uses stdin, stdout and stderr heavily. it also executes a lot of child processes, which also use them. without `poParentStreams`, the child prcesses hang/freeze my whole software. with `poParentStreams`, the child processes stdout and stderr output pollute my main softwares stdout and stderr. both are breaking situations.

**Comment by hoijui** [comment]
finally found a workaround that works for me!

It is essential to close `stdin` and `stderr` before trying to read `stdout`, to prevent hang/freeze:

```nim
let process = osproc.startProcess(
  command = "child-cmd",
  args = [],
  env = nil,
  options = {poUsePath})
process.inputStream.close() # NOTE **Essential** - This prevents hanging/freezing when reading stdout below
process.errorStream.close() # NOTE **Essential** - This prevents hanging/freezing when reading stdout below
let (lines, exCode) = process.readLines()
process.close() # This is probably not required
```
